src/processor/operator/recursive_extend/frontier_scanner.cpp

```diff
@@ -122,12 +122,25 @@ void PathScanner::initDfs(const node_rel_id_t& nodeAndRelID, size_t currentDepth
     initDfs(nbrs->at(0), currentDepth - 1);
 }
 
-static void writePathRels(RecursiveJoinVectors* vectors, sel_t pos, nodeID_t srcNodeID,
-    nodeID_t dstNodeID, relID_t relID, const std::string& labelName) {
-    vectors->pathRelsSrcIDDataVector->setValue<nodeID_t>(pos, srcNodeID);
-    vectors->pathRelsDstIDDataVector->setValue<nodeID_t>(pos, dstNodeID);
-    vectors->pathRelsIDDataVector->setValue<relID_t>(pos, relID);
-    StringVector::addString(vectors->pathRelsLabelDataVector, pos, labelName);
+void PathScanner::writePathNode(idx_t idx, RecursiveJoinVectors& vectors, sel_t vectorPos) {
+    auto nodeID = nodeIDs[idx];
+    vectors.pathNodesIDDataVector->setValue<nodeID_t>(vectorPos, nodeID);
+    auto labelName = tableIDToName.at(nodeID.tableID);
+    StringVector::addString(vectors.pathNodesLabelDataVector, vectorPos, labelName.data(),
+        labelName.length());
+}
+
+void PathScanner::writePathSrcDstNode(idx_t srcNodeIdx, idx_t dstNodeIdx,
+    RecursiveJoinVectors& vectors, sel_t vectorPos) {
+    vectors.pathRelsSrcIDDataVector->setValue<nodeID_t>(vectorPos, nodeIDs[srcNodeIdx]);
+    vectors.pathRelsDstIDDataVector->setValue<nodeID_t>(vectorPos, nodeIDs[dstNodeIdx]);
+}
+
+void PathScanner::writePathRel(common::internalID_t relID, RecursiveJoinVectors& vectors,
+    sel_t vectorPos) {
+    vectors.pathRelsIDDataVector->setValue<relID_t>(vectorPos, relID);
+    StringVector::addString(vectors.pathRelsLabelDataVector, vectorPos,
+        tableIDToName.at(relID.tableID));
 }
 
 void PathScanner::writePathToVector(RecursiveJoinVectors* vectors, sel_t& vectorPos,
```

Comment:
Trivial comment: I think you may want to pass in `RecursiveJoinVectors&` instead of `RecursiveJoinVectors*`.